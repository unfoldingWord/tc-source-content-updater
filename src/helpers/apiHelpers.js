import {delay} from './utils';
import {SORT, STAGE, SUBJECT} from '../index';

const request = require('request');
const DOOR43_CATALOG = `Door43-Catalog`;

/**
 * does http request and returns the response data parsed from JSON
 * @param {string} url
 * @param {number} retries
 * @return {Promise<{Object}>}
 */
export async function makeJsonRequestDetailed(url, retries=5) {
  let result_;
  for (let i = 1; i <= retries; i++) {
    result_ = null;
    try {
      result_ = await new Promise((resolve, reject) => {
        request(url, function(error, response, body) {
          if (error)
            reject(error);
          else if (response.statusCode === 200) {
            let result = body;
            try {
              result = JSON.parse(body);
            } catch (e) {
              reject(e);
            }
            resolve({result, response, body});
          } else {
            reject(`fetch error ${response.statusCode}`);
          }
        });
      });
    } catch (e) {
      if (i >= retries) {
        console.warn('makeJsonRequestDetailed() - error getting manifest data', e);
        throw e;
      }
      result_ = null;
      await delay(500);
      console.log(`makeJsonRequestDetailed() - retry ${i+1} getting manifest data, last error`, e);
    }

    if (result_) {
      break;
    }
  }
  return result_;
}

/**
 * searching subjects
 * @param {array} subjects
 * @param {string} owner
 * @param {number} retries
 * @return {Promise<*>}
 */
async function searchSubjects(subjects, owner, retries=3) {
  const subjectParam = encodeURI(subjects.join(','));
  let fetchUrl = `https://git.door43.org/api/catalog/v3/search?subject=${subjectParam}`;
  if (owner) {
    fetchUrl += fetchUrl + `&owner=${owner}`;
  }
  const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
  return result;
}

/**
 * get published catalog using catalog next v3
 * @return {Promise<*[]>}
 */
export async function getOldCatalogReleases() {
  const released = [];
  const owner = null; // get all owners
  const subjectList = ['Bible', 'Aligned Bible', 'Greek New Testament', 'Hebrew Old Testament', 'Translation Words', 'TSV Translation Notes', 'Translation Academy'];
  // const subjectList = ['Bible', 'Testament', 'Translation Words', 'TSV Translation Notes', 'Translation Academy'];

  try {
    // for (const subject of subjectList) {
    const result = await searchSubjects(subjectList, owner, 5);
    let repos = 0;
    const languages = result && result.languages || [];
    for (const language of languages) {
        const languageId = language.identifier;
        const resources = language.resources || [];
        for (const resource of resources) {
          resource.languageId = languageId;
          resource.resourceId = resource.identifier;
          resource.foundInCatalog = 'OLD';
          resource.full_name = resource.full_name || `${resource.owner}/${resource.repo}`;
          resource.checking_level = resource.checking && resource.checking.checking_level;
          released.push(resource);
          repos++;
        }
      }
      console.log(`has ${repos} items`);
    // }
    console.log(`released catalog has ${released.length} items`);
  } catch (e) {
    console.error('getCatalog() - error getting catalog', e);
    return [];
  }

  return released;
}

/**
 * get published catalog
 * @return {Promise<*[]>}
 */
export async function getCatalog() {
  const catalogReleases = await getOldCatalogReleases();
  const searchParams = {
    subject: SUBJECT.ALL_TC_RESOURCES,
    stage: STAGE.PROD,
  };
  console.log(`found ${catalogReleases.length} items in old catalog`);
  const newCatalogReleases = await searchCatalogNext(searchParams);
  console.log(`found ${newCatalogReleases.length} items in new catalog`);
  // merge catalogs together - catalog new takes precident
  for (const item of newCatalogReleases) {
    const index = catalogReleases.findIndex(oldItem => (item.full_name === oldItem.full_name));
    if (index >= 0) {
      catalogReleases[index] = item; // overwrite item in old catalog
      catalogReleases[index].foundInCatalog = 'NEW+OLD';
    } else {
      catalogReleases.push(item); // add unique item
    }
  }
  console.log(`now ${catalogReleases.length} items in merged catalog`);

  return catalogReleases;
}

/**
 * add parameter to url string
 * @param {*} value
 * @param {object} parameters
 * @param {string} tag
 * @return {*}
 */
function addUrlParameter(value, parameters, tag) {
  if (value) {
    if (parameters) { // see if we need separator
      parameters += '&';
    }
    parameters += `${tag}=${encodeURIComponent(value)}`;
  }
  return parameters;
}

/**
 * filter for supported repos
 * @param {array} resources
 * @return {*[]}
 */
function getCompatibleResourceList(resources) {
  const supported = [];
  for (const item of resources || []) {
    // add fields for backward compatibility
    const languageId = item.language;
    let [, resourceId] = (item.name || '').split(`${languageId}_`);
    resourceId = resourceId || item.name; // if language was not in name, then use name as resource ID
    item.resourceId = resourceId;
    item.languageId = languageId;
    item.checking_level = item.repo && item.repo.checking_level;
    item.foundInCatalog = 'NEW';
    item.modified = item.modified || item.released;

    if (item.zipball_url) {
      item.downloadUrl = item.zipball_url;
    }
    // check for version. if there is one, it will save having to fetch it from DCS later.
    if (item.release) { // if released
      const tagName = item.release.tag_name;
      if (tagName && (tagName[0] === 'v')) {
        item.version = tagName.substr(1);
      }
    } else {
      const branchOrTagName = item.branch_or_tag_name;
      if (branchOrTagName && (branchOrTagName[0] === 'v')) {
        item.version = branchOrTagName.substr(1);
      }
    }
    if (item.subject) {
      item.subject = item.subject.replaceAll(' ', '_');
    }
    // add supported resources to returned list
    if (item.downloadUrl && item.subject && item.name && item.full_name) {
      supported.push(item);
    }
  }
  return supported;
}

/**
 * get published catalog
 * @param {Object} searchParams - details below
 * @param {String} searchParams.owner - if undefined then all are searched
 * @param {String} searchParams.languageId - if undefined then all are searched
 * @param {String} searchParams.subject - one or more separated by comma.  If undefined then all are searched.
 *          Example `Bible,Aligned Bible,Greek New Testament,Hebrew Old Testament,Translation Words,TSV Translation Notes,Translation Academy'
 * @param {Number} searchParams.limit - maximum results to return, default 100
 * @param {Boolean} searchParams.partialMatch - if true will do case insensitive substring matching, default is false
 * @param {String} searchParams.stage - specifies which release stage to be returned out of these stages:
 *                    "prod" - return only the production releases (default)
 *                    "preprod" - return the pre-production release if it exists instead of the production release
 *                    "draft" - return the draft release if it exists instead of pre-production or production release
 *                   "latest" -return the default branch (e.g. master) if it is a valid RC instead of the "prod", "preprod" or "draft".  (default)
 * @param {Number} searchParams.checkingLevel - search only for entries with the given checking level(s). Can be 1, 2 or 3.  Default is any.
 * @param {Number} searchParams.sort - search only for entries with the given checking level(s). Can be 1, 2 or 3.  Default is any.
 * @param {number} retries - number of times to retry calling search API, default 3
 * @return {Promise<*[]|null>}
 */
export async function searchCatalogNext(searchParams, retries=3) {
  let result_ = null;
  const {
    owner,
    languageId,
    subject,
    limit = 100,
    stage = STAGE.LATEST,
    checkingLevel,
    partialMatch,
    sort = SORT.REPO_NAME,
  } = searchParams;

  try {
    let fetchUrl = `https://git.door43.org/api/catalog/v5/search`;
    let parameters = '';
    parameters = addUrlParameter(owner, parameters, 'owner');
    parameters = addUrlParameter(languageId, parameters, 'lang');
    parameters = addUrlParameter(subject, parameters, 'subject');
    parameters = addUrlParameter(limit, parameters, 'limit');
    parameters = addUrlParameter(stage, parameters, 'stage');
    parameters = addUrlParameter(checkingLevel, parameters, 'checkingLevel');
    parameters = addUrlParameter(partialMatch, parameters, 'partialMatch');
    parameters = addUrlParameter(sort, parameters, 'sort');
    if (parameters) {
      fetchUrl += '?' + parameters;
    }
    console.log(`Searching: ${fetchUrl}`);
    const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
    result_ = result && result.data;
  } catch (e) {
    console.warn('searchCatalogNext() - error calling search API', e);
    return null;
  }

  const supported = getCompatibleResourceList(result_);
  return supported;
}

/**
 * does Catalog next API query to get manifest data
 * @param {string} owner
 * @param {string} repo
 * @param {number} retries
 * @return {Promise<{Object}>}
 */
export async function downloadManifestData(owner, repo, retries=5) {
  const fetchUrl = `https://git.door43.org/api/catalog/v5/entry/${owner}/${repo}/master/metadata`;
  try {
    const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
    return result;
  } catch (e) {
    console.warn('getManifestData() - error getting manifest data', e);
    throw e;
  }
}
