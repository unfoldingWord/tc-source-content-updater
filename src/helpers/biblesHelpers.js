/* eslint-disable camelcase */
import fs from 'fs-extra';
import path from 'path-extra';

/**
 * generate manifest.json
 * @param {Object} oldManifest - old manifest data
 * @param {String} RESOURCE_OUTPUT_PATH - folder to store manifest.json
 * @param {Object} resource
 * @return {Object} new manifest data
 */
export function generateBibleManifest(oldManifest, RESOURCE_OUTPUT_PATH, resource) {
  const newManifest = {};
  newManifest.dublin_core = oldManifest.dublin_core; // preserve original manifest data
  newManifest.checking = oldManifest.checking;
  newManifest.projects = oldManifest.projects;
  newManifest.original_manifest = oldManifest;

  // copy some data for more convenient access
  newManifest.language_id = oldManifest.dublin_core.language.identifier;
  newManifest.language_name = oldManifest.dublin_core.language.title;
  newManifest.direction = oldManifest.dublin_core.language.direction;
  newManifest.subject = oldManifest.dublin_core.subject;
  newManifest.resource_id = oldManifest.dublin_core.identifier;
  newManifest.resource_title = oldManifest.dublin_core.title;
  const remoteModifiedTime = resource && resource.remoteModifiedTime || (new Date()).toJSON();
  newManifest.remoteModifiedTime = remoteModifiedTime;
  const oldMainfestIdentifier = oldManifest.dublin_core.identifier
                                      .toLowerCase();
  const identifiers = ['ugnt', 'ubh'];
  newManifest.description = identifiers.includes(oldMainfestIdentifier) ?
    'Original Language' : 'Gateway Language';

  const savePath = path.join(RESOURCE_OUTPUT_PATH, 'manifest.json');
  fs.writeFileSync(savePath, JSON.stringify(newManifest, null, 2));
  return newManifest;
}
