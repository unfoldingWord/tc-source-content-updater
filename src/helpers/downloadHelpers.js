/* eslint-disable curly */
import url from 'url';
import {https, http} from 'follow-redirects';
import fs from 'fs-extra';
import rimraf from 'rimraf';
const HttpAgent = require('agentkeepalive');
const HttpsAgent = require('agentkeepalive').HttpsAgent;

const httpAgent = new HttpAgent();
const httpsAgent = new HttpsAgent();

const MAX_RETRIES = 3;

/**
 * @description Downloads a url to a file.
 * @param {String} uri the uri to download
 * @param {String} dest the file to download the uri to
 * @param {Function} progressCallback receives progress updates
 * @param {number} retries the amount of retries for socket timeouts
 * @return {Promise.<{}|Error>} the status code or an error
 */
export function download(uri, dest, progressCallback, retries = 0) {
  progressCallback = progressCallback || function() {};
  const parsedUrl = url.parse(uri, false, true);
  const makeRequest = parsedUrl.protocol === 'https:' ? https.request.bind(https) : http.request.bind(http);
  const serverPort = parsedUrl.port ? parsedUrl.port : parsedUrl.protocol === 'https:' ? 443 : 80;
  const agent = parsedUrl.protocol === 'https:' ? httpsAgent : httpAgent;
  const file = fs.createWriteStream(dest);
  const options = {
    host: parsedUrl.host,
    path: parsedUrl.path,
    agent: agent,
    port: serverPort,
    method: 'GET',
    timeout: 120000,
  };
  if (process.env.NODE_ENV === 'development') {
    options.rejectUnauthorized = false;
  }

  return new Promise((resolve, reject) => {
    if (!navigator.onLine) {
      reject('Internet offline');
    }

    const req = makeRequest(options, (response) => {
      const size = response.headers['content-length'];
      let progress = 0;

      response.on('data', (chunk) => {
        progress += chunk.length;
        progressCallback(size, progress);
      });

      response.pipe(file);
      file.on('finish', () => {
        fs.exists(dest, (exists) => {
          if (exists) {
            if (response.statusCode === 200) {
              resolve({
                uri,
                dest,
                status: response.statusCode,
              });
            } else {
              req.emit('error', `Download did not succeed, code ${response.statusCode}`);
              fs.removeSync(dest); // remove failed download attempt
            }
          } else {
            req.emit('error', 'Downloaded file does not exist');
          }
        });
      });
    });

    req.on('error', (error) => {
      file.end();
      rimraf.sync(dest);
      req.end();
      if ((retries < MAX_RETRIES) && navigator.onLine) {
        console.warn(`error on resource ${uri} retrying, error: ${error}`);
        setTimeout(() => {
          download(uri, dest, progressCallback, retries + 1).then(resolve).catch(reject);
        }, 500);
      } else {
        console.error(error);
        reject(error);
      }
    });

    req.end();
  });
}
