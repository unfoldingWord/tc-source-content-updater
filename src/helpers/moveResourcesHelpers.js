import fs from 'fs-extra';
import path from 'path';

/**
 * @description transfer an entire resource from source to target directory
 * @param {string} resourceSourcePath - current position of resource
 * @param {string} resourceTargetPath - folder where resources are moved
 * @return {Promise} Move directory Promise
 */
export function moveResources(resourceSourcePath, resourceTargetPath) {
  return new Promise((resolve, reject) => {
    if (resourceSourcePath && resourceSourcePath.length &&
            resourceTargetPath && resourceTargetPath.length &&
            fs.pathExistsSync(resourceSourcePath)) {
      if (fs.pathExistsSync(resourceTargetPath)) {
        fs.removeSync(resourceTargetPath);
      }
      fs.ensureDirSync(path.dirname(resourceTargetPath));
      fs.copySync(resourceSourcePath, resourceTargetPath);
      fs.removeSync(resourceSourcePath);
      resolve(resourceTargetPath);
    } else {
      reject('Invalid parameters to moveResources()');
    }
  });
}

