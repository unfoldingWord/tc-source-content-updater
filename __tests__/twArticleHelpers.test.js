import fs from 'fs-extra';
import path from 'path-extra';
import ospath from 'ospath';
// helpers
import * as twArticleHelpers from '../src/helpers/translationHelps/twArticleHelpers';
import * as resourcesHelpers from '../src/helpers/resourcesHelpers';
// constants
import * as errors from '../src/resources/errors';
import {DOOR43_CATALOG} from '../src/helpers/apiHelpers';

describe('Tests for twArticleHelpers', function() {
  const resource = {
    languageId: 'en',
    resourceId: 'tw',
    version: '8',
    owner: DOOR43_CATALOG,
  };

  beforeEach(() => {
    fs.__resetMockFS();
  });

  it('Test twArticleHelpers.processTranslationWords() for en', () => {
    // given
    const actualExtractedPath = path.join(__dirname, 'fixtures/translationHelps/twExtractedFromCDN');
    const mockedExtractedPath = path.join(ospath.home(), 'translationCore/resources/imports');
    fs.__loadDirIntoMockFs(actualExtractedPath, mockedExtractedPath);
    const outputPath = path.join(ospath.home(), 'translationCore/resources', resource.languageId, 'translationHelps/translationWords', 'v' + resource.version);
    fs.ensureDirSync(outputPath);
    const expectedTypeList = ['kt', 'names', 'other'];
    const expectedKtArticleListLength = 3;
    const expectedNamesArticleListLength = 2;
    const expectedIndexJson = [
      {id: 'apostle', name: 'apostle, apostles, apostleship'},
      {id: 'god', name: 'God'}, {id: 'sanctify', name: 'sanctify, sanctifies, sanctification'}
    ];

    // when
    const result = twArticleHelpers.processTranslationWords(resource, path.join(mockedExtractedPath, resource.languageId + '_' + resource.resourceId), outputPath);
    const indexFile = path.join(outputPath, 'kt', 'index.json');
    const indexJson = fs.readJsonSync(indexFile);
    const typeList = fs.readdirSync(outputPath);
    const ktArticleList = fs.readdirSync(path.join(outputPath, 'kt', 'articles'));
    const namesArticleList = fs.readdirSync(path.join(outputPath, 'names', 'articles'));
    const godFile = path.join(outputPath, 'kt', 'articles', 'god.md');
    const godArticle = fs.readFileSync(godFile, 'utf8');

    // then
    expect(result).toBeTruthy();
    expect(fs.existsSync(indexFile)).toBeTruthy();
    expect(indexJson).toEqual(expectedIndexJson);
    expect(fs.existsSync(godFile)).toBeTruthy();
    expect(typeList).toEqual(expectedTypeList);
    expect(ktArticleList.length).toEqual(expectedKtArticleListLength);
    expect(ktArticleList).toMatchSnapshot();
    expect(godArticle).toMatchSnapshot();
    expect(namesArticleList.length).toEqual(expectedNamesArticleListLength);
    expect(namesArticleList).toMatchSnapshot();
  });

  it('Test twArticlesHelpers.processTranslationWord() for invalid sourcePath', () => {
    // given
    const sourcePath = '/bad/dir';
    const outputPath = path.join(ospath.home(), 'translationCore/resources/imports/en_tw_processed');
    const expectedError = resourcesHelpers.formatError(resource, errors.SOURCE_PATH_NOT_EXIST);

    // when
    expect(() => twArticleHelpers.processTranslationWords(resource, sourcePath, outputPath)).toThrowError(expectedError);
  });

  it('Test twArticlesHelpers.processTranslationWord() for sourcePath not given', () => {
    // given
    const sourcePath = null;
    const outputPath = path.join(ospath.home(), 'translationCore/resources/imports/en_tw_processed');
    const expectedError = resourcesHelpers.formatError(resource, errors.SOURCE_PATH_NOT_GIVEN);

    // when
    expect(() => twArticleHelpers.processTranslationWords(resource, sourcePath, outputPath)).toThrowError(expectedError);
  });

  it('Test twArticlesHelpers.processTranslationWord() forresource not given', () => {
    // given
    const sourcePath = path.join(ospath.home(), 'translationCore/resources/imports/en_tw');
    const outputPath = path.join(ospath.home(), 'translationCore/resources/imports/en_tw_processed');
    const badResource = null;
    fs.ensureDirSync(sourcePath);
    const expectedError = resourcesHelpers.formatError(null, errors.RESOURCE_NOT_GIVEN);

    // when
    expect(() => twArticleHelpers.processTranslationWords(badResource, sourcePath, outputPath)).toThrowError(expectedError);
  });

  it('Test twArticlesHelpers.processTranslationWord() for output path not given', () => {
    // given
    const sourcePath = path.join(ospath.home(), 'translationCore/resources/imports/en_tw');
    const outputPath = null;
    fs.ensureDirSync(sourcePath);
    const expectedError = resourcesHelpers.formatError(resource, errors.OUTPUT_PATH_NOT_GIVEN);

    // when
    expect(() => twArticleHelpers.processTranslationWords(resource, sourcePath, outputPath)).toThrowError(expectedError);
  });
});
