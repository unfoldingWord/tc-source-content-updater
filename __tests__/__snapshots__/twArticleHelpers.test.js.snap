// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for twArticleHelpers Test twArticleHelpers.processTranslationWords() for en 1`] = `
Array [
  "apostle.md",
  "god.md",
  "sanctify.md",
]
`;

exports[`Tests for twArticleHelpers Test twArticleHelpers.processTranslationWords() for en 2`] = `
"# God #

## Facts: ##

In the Bible, the term \\"God\\" refers to the eternal being who created the universe out of nothing. God exists as Father, Son, and Holy Spirit. God's personal name is \\"Yahweh.\\"

* God has always existed; he existed before anything else existed, and he will continue to exist forever.
* He is the only true God and has authority over everything in the universe.
* God is perfectly righteous, infinitely wise, holy, sinless, just, merciful, and loving.
* He is a covenant-keeping God, who always fulfills his promises.
* People were created to worship God and he is the only one they should worship.
* God revealed his name as \\"Yahweh,\\" which means \\"he is\\" or \\"I am\\" or \\"the One who (always) exists.\\"
* The Bible also teaches about false \\"gods,\\" which are nonliving idols that people wrongly worship.

## Translation Suggestions: ##

* Ways to translate \\"God\\" could include \\"Deity\\" or \\"Creator\\" or \\"Supreme Being.\\"
* Other ways to translate \\"God\\" could be \\"Supreme Creator\\" or \\"Infinite Sovereign Lord\\" or \\"Eternal Supreme Being.\\"
* Consider how God is referred to in a local or national language. There may also already be a word for \\"God\\" in the language being translated. If so, it is important to make sure that this word fits the characteristics of the one true God as described above.
* Many languages capitalize the first letter of the word for the one true God, to distinguish it from the word for a false god. 
* Another way to make this distinction would be to use different terms for \\"God\\" and \\"god.\\"
* The phrase \\"I will be their God and they will be my people\\" could also be translated as \\"I, God, will rule over these people and they will worship me.\\"

(Translation suggestions: [How to Translate Names](rc://en/ta/man/translate/translate-names))

(See also: [create](../other/creation.md), [false god](../kt/falsegod.md), [God the Father](../kt/godthefather.md), [Holy Spirit](../kt/holyspirit.md), [false god](../kt/falsegod.md), [Son of God](../kt/sonofgod.md), [Yahweh](../kt/yahweh.md))

## Bible References: ##

* [1 John 01:5-7](rc://en/tn/help/1jn/01/05)
* [1 Samuel 10:7-8](rc://en/tn/help/1sa/10/07)
* [1 Timothy 04:9-10](rc://en/tn/help/1ti/04/09)
* [Colossians 01:15-17](rc://en/tn/help/col/01/15)
* [Deuteronomy 29:14-16](rc://en/tn/help/deu/29/14)
* [Ezra 03:1-2](rc://en/tn/help/ezr/03/01)
* [Genesis 01:1-2](rc://en/tn/help/gen/01/01)
* [Hosea 04:11-12](rc://en/tn/help/hos/04/11)
* [Isaiah 36:6-7](rc://en/tn/help/isa/36/06)
* [James 02:18-20](rc://en/tn/help/jas/02/18)
* [Jeremiah 05:4-6](rc://en/tn/help/jer/05/04)
* [John 01:1-3](rc://en/tn/help/jhn/01/01)
* [Joshua 03:9-11](rc://en/tn/help/jos/03/09)
* [Lamentations 03:40-43](rc://en/tn/help/lam/03/40)
* [Micah 04:4-5](rc://en/tn/help/mic/04/04)
* [Philippians 02:5-8](rc://en/tn/help/php/02/05)
* [Proverbs 24:11-12](rc://en/tn/help/pro/24/11)
* [Psalms 047:8-9](rc://en/tn/help/psa/047/008)

## Examples from the Bible stories: ##

* __[01:01](rc://en/tn/help/obs/01/01)__ __God__  created the universe and everything in it in six days.
* __[01:15](rc://en/tn/help/obs/01/15)__ __God__  made man and woman in his own image.
* __[05:03](rc://en/tn/help/obs/05/03)__ \\"I am __God__  Almighty. I will make a covenant with you.\\"
* __[09:14](rc://en/tn/help/obs/09/14)__ __God__  said, \\"I AM WHO I AM. Tell them, 'I AM has sent me to you.' Also tell them, 'I am Yahweh, the __God__  of your ancestors Abraham, Isaac, and Jacob. This is my name forever.'\\"
* __[10:02](rc://en/tn/help/obs/10/02)__ Through these plagues, __God__  showed Pharaoh that he is more powerful than Pharaoh and all of Egypt's gods.
* __[16:01](rc://en/tn/help/obs/16/01)__ The Israelites began to worship the Canaanite gods instead of Yahweh, the true __God__.
* __[22:07](rc://en/tn/help/obs/22/07)__ You, my son, will be called the prophet of the __Most High God__  who will prepare the people to receive the Messiah!\\"
* __[24:09](rc://en/tn/help/obs/24/09)__ There is only one __God__. But John heard __God__  the Father speak, and saw Jesus the Son and the Holy Spirit when he baptized Jesus.
* __[25:07](rc://en/tn/help/obs/25/07)__ \\"Worship only the Lord your __God__  and only serve him.\\"
* __[28:01](rc://en/tn/help/obs/28/01)__ \\"There is only one who is good, and that is __God__.\\"
* __[49:09](rc://en/tn/help/obs/49/09)__ But __God__  loved everyone in the world so much that he gave his only Son so that whoever believes in Jesus will not be punished for his sins, but will live with __God__  forever.
* __[50:16](rc://en/tn/help/obs/50/16)__ But some day __God__  will create a new heaven and a new earth that will be perfect.

## Word Data: ##

* Strong's: H136, H305, H410, H426, H430, H433, H2486, H2623, H3068, H3069, H3863, H4136, H6697, G112, G516, G932, G935, G1096, G1140, G2098, G2124, G2128, G2150, G2152, G2153, G2299, G2304, G2305, G2312, G2313, G2314, G2315, G2316, G2317, G2318, G2319, G2320, G3361, G3785, G4151, G5207, G5377, G5463, G5537, G5538
"
`;

exports[`Tests for twArticleHelpers Test twArticleHelpers.processTranslationWords() for en 3`] = `
Array [
  "aaron.md",
  "moses.md",
]
`;

exports[`Tests for twArticleHelpers Test twArticleHelpers.processTranslationWords() for en unfoldingWord 1`] = `
Array [
  "apostle.md",
  "god.md",
  "sanctify.md",
]
`;

exports[`Tests for twArticleHelpers Test twArticleHelpers.processTranslationWords() for en unfoldingWord 2`] = `
"# God #

## Facts: ##

In the Bible, the term \\"God\\" refers to the eternal being who created the universe out of nothing. God exists as Father, Son, and Holy Spirit. God's personal name is \\"Yahweh.\\"

* God has always existed; he existed before anything else existed, and he will continue to exist forever.
* He is the only true God and has authority over everything in the universe.
* God is perfectly righteous, infinitely wise, holy, sinless, just, merciful, and loving.
* He is a covenant-keeping God, who always fulfills his promises.
* People were created to worship God and he is the only one they should worship.
* God revealed his name as \\"Yahweh,\\" which means \\"he is\\" or \\"I am\\" or \\"the One who (always) exists.\\"
* The Bible also teaches about false \\"gods,\\" which are nonliving idols that people wrongly worship.

## Translation Suggestions: ##

* Ways to translate \\"God\\" could include \\"Deity\\" or \\"Creator\\" or \\"Supreme Being.\\"
* Other ways to translate \\"God\\" could be \\"Supreme Creator\\" or \\"Infinite Sovereign Lord\\" or \\"Eternal Supreme Being.\\"
* Consider how God is referred to in a local or national language. There may also already be a word for \\"God\\" in the language being translated. If so, it is important to make sure that this word fits the characteristics of the one true God as described above.
* Many languages capitalize the first letter of the word for the one true God, to distinguish it from the word for a false god. 
* Another way to make this distinction would be to use different terms for \\"God\\" and \\"god.\\"
* The phrase \\"I will be their God and they will be my people\\" could also be translated as \\"I, God, will rule over these people and they will worship me.\\"

(Translation suggestions: [How to Translate Names](rc://en/ta/man/translate/translate-names))

(See also: [create](../other/creation.md), [false god](../kt/falsegod.md), [God the Father](../kt/godthefather.md), [Holy Spirit](../kt/holyspirit.md), [false god](../kt/falsegod.md), [Son of God](../kt/sonofgod.md), [Yahweh](../kt/yahweh.md))

## Bible References: ##

* [1 John 01:5-7](rc://en/tn/help/1jn/01/05)
* [1 Samuel 10:7-8](rc://en/tn/help/1sa/10/07)
* [1 Timothy 04:9-10](rc://en/tn/help/1ti/04/09)
* [Colossians 01:15-17](rc://en/tn/help/col/01/15)
* [Deuteronomy 29:14-16](rc://en/tn/help/deu/29/14)
* [Ezra 03:1-2](rc://en/tn/help/ezr/03/01)
* [Genesis 01:1-2](rc://en/tn/help/gen/01/01)
* [Hosea 04:11-12](rc://en/tn/help/hos/04/11)
* [Isaiah 36:6-7](rc://en/tn/help/isa/36/06)
* [James 02:18-20](rc://en/tn/help/jas/02/18)
* [Jeremiah 05:4-6](rc://en/tn/help/jer/05/04)
* [John 01:1-3](rc://en/tn/help/jhn/01/01)
* [Joshua 03:9-11](rc://en/tn/help/jos/03/09)
* [Lamentations 03:40-43](rc://en/tn/help/lam/03/40)
* [Micah 04:4-5](rc://en/tn/help/mic/04/04)
* [Philippians 02:5-8](rc://en/tn/help/php/02/05)
* [Proverbs 24:11-12](rc://en/tn/help/pro/24/11)
* [Psalms 047:8-9](rc://en/tn/help/psa/047/008)

## Examples from the Bible stories: ##

* __[01:01](rc://en/tn/help/obs/01/01)__ __God__  created the universe and everything in it in six days.
* __[01:15](rc://en/tn/help/obs/01/15)__ __God__  made man and woman in his own image.
* __[05:03](rc://en/tn/help/obs/05/03)__ \\"I am __God__  Almighty. I will make a covenant with you.\\"
* __[09:14](rc://en/tn/help/obs/09/14)__ __God__  said, \\"I AM WHO I AM. Tell them, 'I AM has sent me to you.' Also tell them, 'I am Yahweh, the __God__  of your ancestors Abraham, Isaac, and Jacob. This is my name forever.'\\"
* __[10:02](rc://en/tn/help/obs/10/02)__ Through these plagues, __God__  showed Pharaoh that he is more powerful than Pharaoh and all of Egypt's gods.
* __[16:01](rc://en/tn/help/obs/16/01)__ The Israelites began to worship the Canaanite gods instead of Yahweh, the true __God__.
* __[22:07](rc://en/tn/help/obs/22/07)__ You, my son, will be called the prophet of the __Most High God__  who will prepare the people to receive the Messiah!\\"
* __[24:09](rc://en/tn/help/obs/24/09)__ There is only one __God__. But John heard __God__  the Father speak, and saw Jesus the Son and the Holy Spirit when he baptized Jesus.
* __[25:07](rc://en/tn/help/obs/25/07)__ \\"Worship only the Lord your __God__  and only serve him.\\"
* __[28:01](rc://en/tn/help/obs/28/01)__ \\"There is only one who is good, and that is __God__.\\"
* __[49:09](rc://en/tn/help/obs/49/09)__ But __God__  loved everyone in the world so much that he gave his only Son so that whoever believes in Jesus will not be punished for his sins, but will live with __God__  forever.
* __[50:16](rc://en/tn/help/obs/50/16)__ But some day __God__  will create a new heaven and a new earth that will be perfect.

## Word Data: ##

* Strong's: H136, H305, H410, H426, H430, H433, H2486, H2623, H3068, H3069, H3863, H4136, H6697, G112, G516, G932, G935, G1096, G1140, G2098, G2124, G2128, G2150, G2152, G2153, G2299, G2304, G2305, G2312, G2313, G2314, G2315, G2316, G2317, G2318, G2319, G2320, G3361, G3785, G4151, G5207, G5377, G5463, G5537, G5538
"
`;

exports[`Tests for twArticleHelpers Test twArticleHelpers.processTranslationWords() for en unfoldingWord 3`] = `
Array [
  "aaron.md",
  "moses.md",
]
`;
